writer.go

```diff
@@ -117,78 +101,86 @@ type WriterResponseEmitter struct {
 	emitted bool
 }
 
-func (re *WriterResponseEmitter) SetEncoder(mkEnc func(io.Writer) Encoder) {
+func (re *writerResponseEmitter) SetEncoder(mkEnc func(io.Writer) Encoder) {
 	re.enc = mkEnc(re.w)
 }
 
-func (re *WriterResponseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
-	err := re.Emit(&cmdkit.Error{Message: fmt.Sprint(v), Code: errType})
-	if err != nil {
-		panic(err)
+func (re *writerResponseEmitter) CloseWithError(err error) error {
+	cwe, ok := re.c.(interface{ CloseWithError(error) error })
+	if ok {
+		return cwe.CloseWithError(err)
+	}
+
+	if err == nil || err == io.EOF {
+		return re.Close()
 	}
+
+	return errors.New("provided closer does not support CloseWithError")
 }
 
-func (re *WriterResponseEmitter) SetLength(length uint64) {
+func (re *writerResponseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
+	err := &cmdkit.Error{Message: fmt.Sprint(v), Code: errType}
+	log.Errorf("error setting error %q: cannot set error with writerResponseEmitter", err)
+}
+
+func (re *writerResponseEmitter) SetLength(length uint64) {
 	if re.emitted {
 		return
 	}
 
 	*re.length = length
 }
 
-func (re *WriterResponseEmitter) Close() error {
+func (re *writerResponseEmitter) Close() error {
 	return re.c.Close()
 }
 
-func (re *WriterResponseEmitter) Head() Head {
-	return Head{
-		Len: *re.length,
-		Err: re.err,
-	}
-}
-
-func (re *WriterResponseEmitter) Emit(v interface{}) error {
+func (re *writerResponseEmitter) Emit(v interface{}) error {
+	// channel emission iteration
 	if ch, ok := v.(chan interface{}); ok {
 		v = (<-chan interface{})(ch)
 	}
-
 	if ch, isChan := v.(<-chan interface{}); isChan {
-		for v = range ch {
-			err := re.Emit(v)
-			if err != nil {
-				return err
-			}
-		}
-		return nil
+		return EmitChan(re, ch)
 	}
 
-	re.emitted = true
+	// Initially this library allowed commands to return errors by sending an
+	// error value along a stream. We removed that in favour of CloseWithError,
+	// so we want to make sure we catch situations where some code still uses the
+	// old error emitting semantics.
+	// Also errors may occur both as pointers and as plain values, so we need to
+	// check both.
+	debug.AssertNotError(v)
 
 	if _, ok := v.(Single); ok {
 		defer re.Close()
```

Comment:
Will close without error if encoding fails.

Reply:
I don't think I would send the encoding error down the stream. If I try to send an unencodable value through a writer, the stream should live on and I should get an error so I can send something else.

However, we should return the `Close()` error, if there is one.